### PR TITLE
add warmup lr_scheduler

### PR DESCRIPTION
**Pull request by solin319**
Add warmup lr_scheduler, it can overcome optimization challenges early in large batch-size.
This method was used in "Accurate, Large Minibatch SGD: Training ImageNet in 1 Hour".
https://arxiv.org/abs/1706.02677

### CODE REVIEW

**Comment by zhreshold** [comment]
Warmup behavior is nice.
Do you mind adding a combined warmup and multi step scheduler which will be more useful IMO?

**Comment by solin319** [comment]
Yes, I will combine warmup and multi step scheduler in a single class.

**Comment by solin319** [comment]
I have combined warmup and multi step scheduler in a single class.
It can be used as  
“lr_scheduler = mx.lr_scheduler.GradualWarmupWithMultiFactorScheduler(warmup_step=10, begin_lr=0.002, stop_lr=0.1, multifactor_step=[15,18,20], factor=0.1)”

File: python/mxnet/lr_scheduler.py [hunk 1]
```diff
@@ -136,3 +136,80 @@ def __call__(self, num_update):
...
+            Stop updating the learning rate if it is less than this value.
+        multifactor_step: list of int
+            The list of steps to schedule a change
+        factor: float
+            The factor to change the learning rate.    
+        """
+        super(GradualWarmupWithMultiFactorScheduler, self).__init__()
+        if stop_lr <= begin_lr:
+             raise ValueError("stop_lr must larger than begin_lr")
+        assert isinstance(multifactor_step, list) and len(multifactor_step) >= 1
```
Comment (zhreshold) [inline python/mxnet/lr_scheduler.py]: default multifactor_step == int will break, add cast to list first

File: python/mxnet/lr_scheduler.py [hunk 2]
```diff
@@ -136,3 +136,80 @@ def __call__(self, num_update):
...
+        for i, _step in enumerate(multifactor_step):
+            if i != 0 and multifactor_step[i] <= multifactor_step[i-1]:
+                raise ValueError("Schedule step must be an increasing integer list")
+            if _step < 1:
+                raise ValueError("Schedule step must be greater or equal than 1 round")
+        if multifactor_step[0] <= warmup_step:
```
Comment (zhreshold) [inline python/mxnet/lr_scheduler.py]: default step 100 will break here too

**Comment by piiswrong** [comment]
Please add test.

Can this be added as an additional argument of MultiFactorScheduler?

**Comment by solin319** [comment]
I have added  additional argument in MultiFactorScheduler class.
Warmup can be used when 'warmup_step' > 1.

I added a test file in test/commom/test_lr_sceduler.py.
Following is a test result
```
python test_lr_scheduler.py --lr-factor=0.1 --warmup_step=5 --lr-steps='8,10'

INFO:root:step 1 lr = 0.020000
INFO:root:step 2 lr = 0.040000
INFO:root:step 3 lr = 0.060000
INFO:root:step 4 lr = 0.080000
INFO:root:Update[5]: now learning rate arrived at 1.00000e-01, will not warm up in the future
INFO:root:step 5 lr = 0.100000
INFO:root:step 6 lr = 0.100000
INFO:root:step 7 lr = 0.100000
INFO:root:step 8 lr = 0.100000
INFO:root:Update[9]: Change learning rate to 1.00000e-02
INFO:root:step 9 lr = 0.010000
INFO:root:step 10 lr = 0.010000
INFO:root:Update[11]: Change learning rate to 1.00000e-03
INFO:root:step 11 lr = 0.001000
INFO:root:step 12 lr = 0.001000
INFO:root:step 13 lr = 0.001000
INFO:root:step 14 lr = 0.001000

```

**Comment by solin319** [comment]
I have moved test_lr_scheduler in tests/python/unittest/test_lr_scheduler.py.
I  have removed argparse and added some test cases in the test file.
@zhreshold

File: tests/python/unittest/test_lr_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        logging.info("step %d lr = %f", i, cur_lr)
+
+    if warmup_step > 1:
+        assert mx.test_utils.almost_equal(res_lr[warmup_step], warmup_lr, 1e-10) 
+        lr = warmup_lr
+    for i in range(len(steps)):
+        assert mx.test_utils.almost_equal(res_lr[steps[i]], lr * pow(lr_factor, i + 1), 1e-10)    
+
+if __name__ == "__main__":
+    #Legal input
```
Comment (zhreshold) [inline tests/python/unittest/test_lr_scheduler.py]: Wrap all these tests in a function like test_multi_lr_scheduler(), and use 
```
if __name__ == '__main__':
    import nose
    nose.runmodule()

# or maybe without nose
if __name__ == '__main__':
    test_multi_step_lr_scheduler()
```
to trigger all tests, we can have more unit test for lr_schedulers in the future.

File: tests/python/unittest/test_lr_scheduler.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    test_lr_sceduler(lr = 0.2, steps = [8,12], lr_factor = 0.1, warmup_step = 0, warmup_lr = 0.1)
+    test_lr_sceduler(lr = 0.02, steps = [8,12], lr_factor = 0.1, warmup_step = 1, warmup_lr = 0.1)
+    test_lr_sceduler(lr = 0.02, steps = [8,12], lr_factor = 0.3, warmup_step = 5, warmup_lr = 0.1)
+    test_lr_sceduler(lr = 0.002, steps = [8,12], lr_factor = 0.1, warmup_step = 7, warmup_lr = 0.1)
```
Comment (zhreshold) [inline tests/python/unittest/test_lr_scheduler.py]: add a default test like: test_lr_sceduler(steps=[100, 200]) to test the default parameters if user don't specify anything.

**Comment by solin319** [comment]
I have used 'nose' to test and add default test 'multi_lr_sceduler(lr = 0.02, steps=[100, 200])'.
@zhreshold

**Comment by zhreshold** [comment]
please `make pylint` at root, fix potential error and wait for CI to finish.

**Comment by solin319** [comment]
Add pylint and license as "test_operator.py"

**Comment by zhreshold** [comment]
You are all set. Can you rebase master and push again?, the checks failed with other problems.

**Comment by solin319** [comment]
move to #8107
